### PR TITLE
Track social media clicks

### PR DESCRIPTION
**Pull request by lizconlan**
Creates a Google Analytics event whenever someone exits the site using a social media link (Experiment 71)

Events reporting example:

screen shot 2016-03-18 at 15 22 28

Also adds a new helper for constructing the event code string to go in the `onclick` attribute

### CODE REVIEW

File: app/helpers/link_to_helper.rb [hunk 1]
```diff
@@ -254,6 +254,20 @@ def unhappy_url(info_request = nil)
     end
   end
 
+  # Google Analytics
+
+  # if used, value must an integer
+  def analytics_event_code(category, action, label=nil, value=nil, label_is_script=false)
```
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: An options hash would be much clearer here. This could also do with some [documentation](http://tomdoc.org/) about what the arguments do. I wouldn't know how to use this by only looking at the method signature.

File: app/helpers/link_to_helper.rb [hunk 2]
```diff
@@ -254,6 +254,20 @@ def unhappy_url(info_request = nil)
...
+    value = 1 unless value
```
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: > `# if used, value must an integer`

Lets handle this:

```ruby
irb(main):006:0> Integer(1)
=> 1
irb(main):004:0> Integer('1')
=> 1
irb(main):005:0> Integer('Word')
ArgumentError: invalid value for Integer(): "Word"
        from (irb):5:in `Integer'
        from (irb):5
        from /usr/bin/irb:12:in `<main>'
```

Comment (garethrees) [inline app/helpers/link_to_helper.rb]: I'd probably do something like:

begin
  value = Integer(value)
rescue ArgumentError
  raise ArgumentError, %Q(:value option must be an Integer: "#{ options[:value] }")
end

File: app/helpers/link_to_helper.rb (hunk 1 above)
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: I'd name the required arguments `event_category` and `event_action` so that they map to the titles used [by Google](https://github.com/mysociety/alaveteli/pull/3110#issue-141910287)

File: app/helpers/link_to_helper.rb [hunk 3]
```diff
@@ -254,6 +254,20 @@ def unhappy_url(info_request = nil)
...
+    label_string = if label
+      ",#{format_event_label(label, label_is_script)},#{value}"
+    else
+      ""
```
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: Don't need the else, since `nil` will interpolate as an empty string

File: app/helpers/link_to_helper.rb [hunk 4]
```diff
@@ -273,4 +287,12 @@ def message_path(message)
     message_url(message, :only_path => true)
   end
 
+  def format_event_label(label, is_script)
```
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: It would be worth adding a line of documentation about why you might need to use this.

File: spec/helpers/link_to_helper_spec.rb [hunk 5]
```diff
@@ -146,4 +146,32 @@
 
   end
 
+  describe "when constructing a google analytics event string" do
```
Comment (garethrees) [inline spec/helpers/link_to_helper_spec.rb]: Much easier to find the spec you're after when the description is the method name:

```diff
- describe "when constructing a google analytics event string" do
+ describe '#analytics_event_code' do
```

File: spec/helpers/link_to_helper_spec.rb [hunk 6]
```diff
@@ -146,4 +146,32 @@
...
+
+      it "uses the supplied value if there is one" do
+        expect(analytics_event_code("test category", "test", "test label", 42)).to eq(
+          "if (pageTracker) { pageTracker._trackEvent('test category','test','test label',42) };"
+        )
+      end
+
+      it "doesn't wrap the label in quotes if passed label_is_script=true" do
+        expect(analytics_event_code("test category", "test", "window.location.href", 42, true)).to eq(
+          "if (pageTracker) { pageTracker._trackEvent('test category','test',window.location.href,42) };"
```
Comment (garethrees) [inline spec/helpers/link_to_helper_spec.rb]: Why might I want to use `window.location.href` as the 3rd param vs `'test label'` as in the previous spec above? Presumably the former is going to dynamically grab the window URL; Is it valid to supply a String?

File: app/helpers/link_to_helper.rb (hunk 1 above)
Comment (garethrees) [inline app/helpers/link_to_helper.rb]: I have a mild preference for calling this method `track_analytics_event` as its a bit more "Tell, don't ask" (okay, not exactly, but I think better describes what will _happen_ when used, rather than what it _is_). Shrug! :lollipop:

File: spec/helpers/link_to_helper_spec.rb (hunk 6 above)
Comment (lizconlan) [inline spec/helpers/link_to_helper_spec.rb]: The use case for this is the widget code - which does something like `window.top.location.href` which tells us where the code's been embedded, not essential but a nice to have. If we wrap it as a String, it will literally log "window.top.location.href" which isn't what we want to capture (I have accidentally proved this a couple of times).

So yes, depending on the scenario, if we want to use the optional label/value feature we can supply a we can supply a String literal if that's useful or some javascript so that the browser assembles the value that the `onclick` event uses instead.

(or to put it another way... I agree, I need to write some helpful documentation!)

File: app/helpers/link_to_helper.rb (hunk 2 above)
Comment (lizconlan) [inline app/helpers/link_to_helper.rb]: note to self: proposed new method signature:

`def analytics_event_code(event_category, event_action, options={})`

(original method name used for readability purposes)

**Comment by garethrees** [comment]
The general gist of this is good, with some comments in line.

I'm concerned that allowing "any old strings" to the tracking method will result in a complete mess of duplicated values over time (e.g. someone may use `'Facebook exit'` when they mean `'Facebook Page Exit'`). @lizconlan commented that this has already happened during testing.

[Google's advice](https://support.google.com/analytics/answer/1033068?hl=en) is pretty clear on this, and cites a similar example:

>  For example, if you plan to call your video tracking category "Video" and later forget and use the plural "Videos," you will have two separate categories for video tracking.

TL;DR: Put some thought in to your naming conventions.

My thinking is that we should have some sort of "dictionary" of allowed trackable actions.

Working backwards, I think we want to be passing a value object in to the helper:

```ruby
@event = AnalyticsEvent.new(:category => 'C', :action => 'A', :label => 'L', :value => 1, :label_is_script => false)
```
```erb
<%= track_analytics_event(@event) %>
```

So about this dictionary…

I think the interface should be specifying a _Thing_, rather than a `String`.

My initial thinking was something like:

```ruby
module AnalyticsEvent
  module Facebook
    module PageExit
      SHARE_REQUEST = 'Share Request'
      SHARE_BATCH_REQUEST = 'Share Batch Request'
    end
  end
end

# Convert data from constant to an AnalyticsEvent instance under the hood?
track_analytics_event(AnalyticsEvent::Facebook::PageExit::SHARE_REQUEST)
```

I like the idea of modules and constants because its easy for a theme to define additional tracking values. The problem here is that `SHARE_REQUEST` is a String, so we lose the category/label/value attributes. We could make it a `Hash`, but then we're duplicating information that the module is giving us (`PageExit` in this case might be a category). We could have a giant `Hash` (`AnalyticsEvent::ALAVETELI_EVENTS`) but that _should_ be frozen, so adding to it is problematic.

I think the real problem here is that we don't have any schema for our event tracking (<q>Determine in advance all elements for which you want to track data – https://goo.gl/vq8I63</q>) so its pretty hard to model that in code.

I don't think its the _end of the world_ if we just roll with supplying strings, but we need to be happy that we're making a conscious decision to introduce technical debt, and if we're doing that we should have a plan for what we're going to do to repay it.

**Comment by lizconlan** [comment]
> I'm concerned that allowing "any old strings" to the tracking method will result in a complete mess of duplicated values over time (e.g. someone may use 'Facebook exit' when they mean 'Facebook Page Exit'). @lizconlan commented that this has already happened during testing.

Worse, it's on the live site, see: https://github.com/mysociety/alaveteli/pull/3110/commits/ea9390d8f0a10904504265d297017409dae034b4#diff-6afc1c2b49c42f3d28d7a868fe7b43a5L52 / https://github.com/mysociety/alaveteli/pull/3110/commits/ea9390d8f0a10904504265d297017409dae034b4#diff-6afc1c2b49c42f3d28d7a868fe7b43a5L57

(I'm more than happy to pull these out into a separate PR/ticket so that we can treat fixing that as a separate issue)

**Comment by lizconlan** [comment]
> @event = AnalyticsEvent.new(:category => 'C', :action => 'A', :label => 'L', :value => 1, :label_is_script => false)

Torn, the initial reason for wanting to have `category` and `action` as params was to bake into the method signature that those 2 params are mandatory whereas the rest are optional extras. But I like the object creation syntax.

**Comment by garethrees** [comment]
> Torn, the initial reason for wanting to have `category` and `action` as params was to bake into the method signature that those 2 params are mandatory whereas the rest are optional extras.

They could totally be optional within the value object: `AnalyticsEvent.new(:category => 'C', :action => 'A')`.

**Comment by garethrees** [comment]
Maybe a better idea for a "dictionary" is to actually map directly to the attributes:

```ruby
module AnalyticsEvent
  module Category
    # "Code" name = "Human" name
    OUTBOUND = 'Outbound'
  end

  module Action
    FACEBOOK_EXIT = 'Facebook Exit'
    TWITTER_EXIT = 'Twitter Exit'
  end
end
```

This improves things slightly:

```erb
<%= track_analytics_event(AnalyticsEvent::Category::OUTBOUND, AnalyticsEvent::Action::FACEBOOK_EXIT) %>
```

**Comment by lizconlan** [comment]
>> Torn, the initial reason for wanting to have `category` and `action` as params was to bake into the method signature that those 2 params are mandatory whereas the rest are optional extras.
>
> They could totally be optional within the value object: `AnalyticsEvent.new(:category => 'C', :action => 'A')`.

Not immediately obvious that `category` and `action` are required though (raising an `ArgumentError` or similar would help, of course)

**Comment by garethrees** [comment]
> Not immediately obvious that `category` and `action` are required though (raising an `ArgumentError` or similar would help, of course)

Ah, with you. Can make those arguments of a value object and use option keys for the other parts.

Note there's also http://api.rubyonrails.org/v3.2.22/classes/Hash.html#method-i-assert_valid_keys, but yeah, I think I prefer the method signature option.

**Comment by lizconlan** [comment]
Right so we need...

- [ ] An opinionated "model" for tracking this data (and a place in the code for it to live)
- [ ] Some library/helper code to make constructing the string to put in the `onclick` attribute simpler / less error prone for the outbound link tracking
- [ ] To use this to track outbound links
- [ ] A fix for "widget" vs "widgets" as a category before we collect a ton of data which is awkward to use

**Comment by garethrees** [comment]
> Some library/helper code to make constructing the string to put in the onclick attribute simpler / less error prone for the outbound link tracking

I'm actually thinking maybe not bother with the intermediary object, given that [this solution](https://github.com/mysociety/alaveteli/pull/3110#issuecomment-199881842) no longer uses magic strings.

My remaining concern with the above is that it would still be possible to provide an "invalid" combination, e.g.

```erb
<%= track_analytics_event(AnalyticsEvent::Category::INBOUND, 
                          AnalyticsEvent::Action::FACEBOOK_EXIT) %>
```

Maybe that's okay? I think its preferable to both the existing implementation (magic strings) and using an intermediary value object that complicates "just tracking the damn event".

**Comment by lizconlan** [comment]
\*starts in the middle\*

How about:

A dedicated helper for outbound link tracking:
```
analytics_track_outbound_link_event(
  AnalyticsEvent::Action::FACEBOOK_EXIT,
  {:label => "Footer link"})
```
(clunky method name, counter-suggestions welcome)

As well as exposing the generic method (which this would call)

**Comment by garethrees** [comment]
Yeah, something like that would work, but y'know what; I think that might be abstracting too soon. We only have a handful of events defined so I don't think its the end of the world to use the "generic" method for everything. Abstracting later is easier once we've got the constants sending the same values to google, so changing the methods called can be done later.

**Comment by lizconlan** [comment]
At the risk of over-complicating things, I have half an idea about making Events part of Categories so that it's either immediately apparent that something odd is going on if we end up calling

```erb
<%= track_analytics_event(AnalyticsEvent::Category::INBOUND, 
                          AnalyticsEvent::Category::OUTBOUND::EXIT_ACTION) %>
```

and could even raise an error if the Action is not in an allowable choice for the category

then maybe we could have a simplified version of the call so that we just pass the Action and use the parent category by default

(could well be more trouble than it's worth tho)

**Comment by garethrees** [comment]
> At the risk of over-complicating things, I have half an idea about making Actions part of Categories

I don't think we really have a good enough idea of how all these map to each other at the moment. My main concern here is that we have a dictionary of known names, rather than having to grep the entire codebase next time we want to add a tracking event. Honestly, I think the right solution for now is to just have a [module per component](https://github.com/mysociety/alaveteli/pull/3110#issuecomment-199881842).

Sorry for any confusion around this. I think the problem here was that we didn't have a ticket that set out what we were trying to do, so as _my_ understanding (as a reviewer) of that improved during discussions around the initial review, my ideas changed here, too.

File: app/helpers/analytics_helper.rb [hunk 7]
```diff
@@ -20,7 +20,7 @@ def track_analytics_event(event_category, event_action, options={})
       label_string = ",#{format_event_label(label, label_is_script)},#{value}"
     end
     event_args = "'#{event_category}','#{event_action}'#{label_string}"
-    "if (pageTracker) { pageTracker._trackEvent(#{event_args}) };"
+    "if (ga) { ga('send','event',#{event_args}) };"
```
Comment (garethrees) [inline app/helpers/analytics_helper.rb]: Just noting to myself that https://github.com/mysociety/alaveteli/pull/3160 is the PR for #3107.

File: spec/helpers/analytics_helper_spec.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    context "when passed a label" do
+      it "includes the default value 1 if none is supplied" do
+        expect(track_analytics_event("test category", "test", :label => "test label")).to eq(
+          "if (ga) { ga('send','event','test category','test','test label',1) };"
+        )
+      end
+
+      it "uses the supplied value if there is one" do
+        expect(track_analytics_event(
+          "test category", "test", {:label => "test label", :value => 42}
```
Comment (garethrees) [inline spec/helpers/analytics_helper_spec.rb]: Don't need the braces around the options:

```ruby
track_analytics_event "test category", "test", :label => "test label", :value => 42
# => "if (ga) { ga('send','event','test category','test','test label',42) };"
```

File: spec/helpers/analytics_helper_spec.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        )).to eq(
+          "if (ga) { ga('send','event','test category','test','test label',42) };"
+        )
+      end
+
+      it "doesn't wrap the label in quotes if passed label_is_script=true" do
+        expect(track_analytics_event(
+          "test category",
+          "test",
+          {:label => "window.location.href", :value => 42, :label_is_script => true}
```
Comment (garethrees) [inline spec/helpers/analytics_helper_spec.rb]: Don't need the braces around the options

File: spec/helpers/analytics_helper_spec.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        )).to eq(
+          "if (ga) { ga('send','event','test category','test',window.location.href,42) };"
+        )
+      end
+
+      it "raises an ArgumentError if value is not nil or an Integer" do
```
Comment (garethrees) [inline spec/helpers/analytics_helper_spec.rb]: "is not nil"?

File: app/views/widgets/show.html.erb [hunk 11]
```diff
@@ -37,24 +37,40 @@
       <% elsif @existing_track %>
         <%= link_to update_url(:track_id => @existing_track.id, :track_medium => "delete", :r => show_user_profile_url(:url_name => @user.url_name)), :target => '_top',
           :class => 'alaveteli-widget__bottom__action alaveteli-widget__button--unsubscribe',
-          :onclick => "if (pageTracker) { pageTracker._trackEvent('widgets', 'unsubscribe', window.top.location.href, 1)};" do %>
+          :onclick => "#{track_analytics_event(
```
Comment (garethrees) [inline app/views/widgets/show.html.erb]: Not sure why the helper calls in this file are interpolated where others aren't.

Can also drop the braces on the options hash.

File: app/views/widgets/show.html.erb [hunk 12]
```diff
@@ -37,24 +37,40 @@
...
+                          AnalyticsEvent::Category::WIDGET_CLICK,
+                          AnalyticsEvent::Action::WIDGET_VOTE,
+                          {:label => "window.top.location.href",
+                           :label_is_script => true})}" do %>
           <%= _('I also want to know!') %>
         <% end %>
       <% else %>
         <%= form_tag request_widget_votes_url(@info_request), :method => 'post', :target => '_top' do %>
-          <%= submit_tag _('I also want to know!'), :class => 'alaveteli-widget__bottom__action alaveteli-widget__button--create-vote', :onclick => 'if (pageTracker) { pageTracker._trackEvent("widget", "vote", window.top.location.href, 1)};' %>
+          <%= submit_tag _('I also want to know!'), :class => 'alaveteli-widget__bottom__action alaveteli-widget__button--create-vote', :onclick => '#{track_analytics_event(
```
Comment (garethrees) [inline app/views/widgets/show.html.erb]: You can't interpolate in single quotes.

File: app/helpers/analytics_helper.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
+# -*- encoding : utf-8 -*-
+module AnalyticsHelper
+
+  # helpers for embedding Google Analytics code
+
+  def track_analytics_event(event_category, event_action, options={})
```
Comment (garethrees) [inline app/helpers/analytics_helper.rb]: As mentioned in https://github.com/mysociety/alaveteli/pull/3110#discussion_r56967008, it would be good to get some  [documentation](http://tomdoc.org/) about what the arguments do, particularly mentioning where to find the established categories and actions (i.e. `AnalyticsEvent::Category` and `AnalyticsEvent::Action`)

File: spec/helpers/analytics_helper_spec.rb (hunk 10 above)
Comment (garethrees) [inline spec/helpers/analytics_helper_spec.rb]: Just to clarify, I'm not sure what the intent of "is not nil" was here.

Given the expectation in this spec, I'd expect its description to read: `it "raises an ArgumentError if the :value option is not an Integer"`.

Should we be testing for some `nil` case in a new spec?

**Comment by garethrees** [comment]
Needs a bit of extra work, and is waiting on https://github.com/mysociety/alaveteli/pull/3160 before it can be merged anyway. Also needs a changelog entry.

File: lib/analytics_event.rb [hunk 14]
```diff
@@ -13,7 +13,7 @@ module Action
     TWITTER_EXIT = 'Twitter Exit'
     WORDPRESS_EXIT = 'WordPress Exit'
     WIDGET_VOTE = 'Vote'
-    WIDGET_SIGNIN = 'Sign in to vote'
+    WIDGET_SIGNIN = 'Sign in to track'
```
Comment (lizconlan) [inline lib/analytics_event.rb]: better/more accurate summary of click intention - from https://github.com/mysociety/alaveteli/pull/3160#discussion_r61054212

File: spec/helpers/analytics_helper_spec.rb (hunk 10 above)
Comment (lizconlan) [inline spec/helpers/analytics_helper_spec.rb]: the valid values are Integer or nil/no value &lt;/not_helping&gt;

Comment (lizconlan) [inline spec/helpers/analytics_helper_spec.rb]: (rewriting tests)

File: app/views/widgets/show.html.erb (hunk 11 above)
Comment (lizconlan) [inline app/views/widgets/show.html.erb]: \*sigh\* overly literal thinking, of course I don't have to tell `:onclick =>` to wrap the results of the method call in quotes, the Rails helper will sort that out for itself - fixed in d19479d

**Comment by lizconlan** [comment]
Stalling on the CHANGELOG for the moment, will add when the code feels less subject to change.

**Comment by garethrees** [comment]
👍
